### PR TITLE
GraphQL filter by quarkus version

### PR DESCRIPTION
**Pull request by pjgg**
- Update readme with previous Quarkus extension Version changes.
- Update GraphQL `repositoriesByExtensions` in order to search by extension name and version
- Update GraphQL `repositories`  in order to filter by quarkusVersion
- Create GraphQL `repositoriesByExtensionsArtifactIds` that corresponds with the old `repositoriesByExtensions` operation

### CODE REVIEW

File: rest-api/src/main/java/io/quarkus/qe/SearchResource.java [hunk 1]
```diff
@@ -26,19 +29,19 @@
...
     @Description("Get repository by repository URL")
-    public Repository getRepositoryByRepoUrl(String repoUrl) throws RepositoryNotFoundException {
+    public Repository getRepositoryByRepoUrl(@Name("repoUrl") String repoUrl) throws RepositoryNotFoundException {
         return repositoryService.findByRepoUrl(repoUrl);
     }
 
     @Query("repositoriesByExtensions")
     @Description("Get repositories using a list of extensions")
-    public List<Repository> getRepositoriesByExtension(List<String> extensions) {
+    public List<Repository> getRepositoriesByExtension(@Name("extensions") List<ExtensionsQuery> extensions) {
```
Comment (Sgitario) [inline rest-api/src/main/java/io/quarkus/qe/SearchResource.java]: Why not using directly QuarkusExtension? We can use the hibernate constraints only for the extension name.

File: README.md [hunk 2]
```diff
@@ -164,16 +184,35 @@ At the moment, there are four queries:
...
-   }
+  }
+}
+```
+
+- repositories that use any of a list of artifactIds:
+
+```
+{
+  repositoriesByExtensionsArtifactIds(artifactIds: ["quarkus-grpc", "quarkus-jdbc-h2"]) {
```
Comment (Sgitario) [inline README.md:205]: What is the difference between repositoriesByExtensionsArtifactIds and repositoriesByExtensions ? As far I see, the only difference is that this query accepts a list of string and the latter a list of extensions which is more usable and powerful.

File: rest-api/src/main/java/io/quarkus/qe/SearchResource.java [hunk 3]
```diff
@@ -20,25 +22,31 @@
 
     @Query("repositories")
     @Description("Get all repositories")
-    public List<Repository> getAllRepositories() {
-        return repositoryService.findAll();
+    public List<Repository> getAllRepositories(@Name("quarkusVersions") @DefaultValue("[]") List<String> versions) {
```
Comment (Sgitario) [inline rest-api/src/main/java/io/quarkus/qe/SearchResource.java:25]: Nice!

File: rest-api/src/main/java/io/quarkus/qe/dto/ExtensionsQuery.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    private static final String REQUIRED_NAME_MSG = "Missing required Artifact id.";
+    private static final String REQUIRED_VERSION_MSG = "Missing required Artifact version.";
+
+    @NotEmpty(message = REQUIRED_NAME_MSG)
+    @NotNull(message = REQUIRED_NAME_MSG)
+    public String name;
+
+    @NotEmpty(message = REQUIRED_VERSION_MSG)
+    @NotNull(message = REQUIRED_VERSION_MSG)
```
Comment (Sgitario) [inline rest-api/src/main/java/io/quarkus/qe/dto/ExtensionsQuery.java]: The version of an extension can be null.

File: rest-api/src/main/java/io/quarkus/qe/services/RepositoryService.java [hunk 5]
```diff
@@ -40,12 +42,37 @@ public Repository findById(Long id) throws RepositoryNotFoundException {
...
+                sql.append(String.format(nameVersionStatement, extensionVersion.name, extensionVersion.version));
+            }
+        }
+
+        // make query
+        return RepositoryEntity.<RepositoryEntity> find(sql.toString())
+                .stream()
                 .map(repositoryMarshaller::fromEntity)
                 .collect(Collectors.toList());
+
```
Comment (Sgitario) [inline rest-api/src/main/java/io/quarkus/qe/services/RepositoryService.java]: Extra space here.

File: rest-api/src/main/java/io/quarkus/qe/services/RepositoryService.java [hunk 6]
```diff
@@ -57,8 +84,17 @@ public Repository findByRepoUrl(String repoUrl) throws RepositoryNotFoundExcepti
...
+        StringBuilder sql = new StringBuilder();
+        String quarkusVersionParamName = "quarkusVersion";
+        sql.append("select distinct repo from repository repo\t");
+        Map<String, Object> versionsParam = getQuarkusVersionAsQueryParam(quarkusVersionParamName, versions);
+
+        if (!versionsParam.isEmpty()) {
+            sql.append("where quarkusVersion IN :" + quarkusVersionParamName);
+        }
+
+        return RepositoryEntity.<RepositoryEntity> find(sql.toString(), versionsParam)
```
Comment (Sgitario) [inline rest-api/src/main/java/io/quarkus/qe/services/RepositoryService.java:97]: Can you refactor this piece of code to use Parameters instead of maps? https://quarkus.io/guides/hibernate-orm-panache#query-parameters

File: rest-api/src/test/java/io/quarkus/qe/SearchResourceTest.java [hunk 7]
```diff
@@ -66,17 +80,31 @@ public void testRepositoryByUrlQuery() {
...
+        givenExistingRepositoryWithExtensions("quarkus-a", "quarkus-b");
+        whenRunGraphqlRepositoriesByExtensionsArtifactIdQuery("quarkus-not-found");
+        thenRepositoryIsNotFound();
+    }
+
     @Test
     public void testRepositoriesByExtensionsQuery() {
         givenExistingRepositoryWithExtensions("quarkus-a", "quarkus-b");
-        whenRunGraphqlRepositoriesByExtensionsQuery("quarkus-b");
+        whenRunGraphqlRepositoriesByExtensionsQuery("quarkus-b", "quarkus-a");
```
Comment (Sgitario) [inline rest-api/src/test/java/io/quarkus/qe/SearchResourceTest.java]: why did you need this change?

File: rest-api/src/test/java/io/quarkus/qe/SearchResourceTest.java [hunk 8]
```diff
@@ -52,6 +52,20 @@ public void testRepositoriesQuery() {
         thenRepositoryIsFound();
     }
 
+    @Test
+    public void testRepositoriesQueryFilterByQuarkusVersion() {
+        givenExistingRepository();
+        whenRunGraphqlRepositoriesQuery(entity.quarkusVersion.id);
```
Comment (Sgitario) [inline rest-api/src/test/java/io/quarkus/qe/SearchResourceTest.java]: Entity is a test field, add a new whenRunGraphqlRepositoriesQuery with zero arguments to improve readability. Actually, the version parameter should be optional.

**Review by Sgitario — CHANGES_REQUESTED** [review]
I don't see why we need two queries repositoriesByExtensionsArtifactIds and repositoriesByExtensions for getting the extensions... As far I see, the only difference is that this query accepts a list of string and the latter a list of extensions which is more usable and powerful.

File: README.md (hunk 2 above)
Comment (pjgg) [reply README.md:205]: repositoriesByExtensions -> returns all the repos that contain any of the given extensions (artifactsID + Version).

repositoriesByExtensionsArtifactIds -> returns all the repos that contain any of the artifacts (you don't care about the version). So if you don't really know the version that you are looking for, this is a better query.

File: rest-api/src/main/java/io/quarkus/qe/SearchResource.java (hunk 1 above)
Comment (pjgg) [reply rest-api/src/main/java/io/quarkus/qe/SearchResource.java]: Because I think that is not a very good idea to push our model to an endpoint or a database. I mean, your model must be decoupled to any port or layer. Otherwise if the model change, we will break the contract (API).

Comment (Sgitario) [reply rest-api/src/main/java/io/quarkus/qe/SearchResource.java]: This model is intended to be shared. Indeed, our REST API uses it.

File: README.md (hunk 2 above)
Comment (Sgitario) [reply README.md:205]: My point is that why not using directly the first query. If the version is not provided, it will behave as the second query. Still don't see the point of having two queries.

**Review by Sgitario — APPROVED** [review]
